luarules/gadgets/game_unit_sharing_mode.lua

```diff
@@ -0,0 +1,101 @@
+local gadget = gadget ---@type Gadget
+
+function gadget:GetInfo()
+	return {
+		name    = 'Unit Sharing Control',
+		desc    = 'Controls unit sharing based on modoption settings',
+		author  = 'Rimilel',
+		date    = 'May 2024 / April 2025',
+		license = 'GNU GPL, v2 or later',
+		layer   = 0,
+		enabled = true
+	}
+end
+
+----------------------------------------------------------------
+-- Synced only
+----------------------------------------------------------------
+if not gadgetHandler:IsSyncedCode() then
+	return false
+end
+
+local unitSharingMode = Spring.GetModOptions().unit_sharing_mode or "enabled"
+local unitMarketEnabled = Spring.GetModOptions().unit_market or false
+
+-- Disable the gadget only if unit sharing is fully enabled
+if unitSharingMode == "enabled" then
+	return false
+end
+
+--[[ Engine Callin Note: AllowUnitTransfer blocks if *any* gadget returns false.
+     This gadget must therefore explicitly check for and allow certain transfer types
+     (Market, /take, Capture) to prevent this gadget's rules (disabled/t2cons)
+     from blocking valid engine or other gadget actions.
+]]
+
+-- Flags for specific transfer types that should cleanly bypass this gadget's sharing rules.
+local transferOverrideFlags = {
+	"marketTransferInProgress",
+	-- Note: takeCommand logic uses a separate check function due to engine limitations.
+}
+
+local function isT2Constructor(unitDef)
+	if not unitDef then return false end
+
+	return not unitDef.isFactory
+			and #(unitDef.buildOptions or {}) > 0
+			and unitDef.customParams.techlevel == "2"
+end
+
+-- Handles the specific condition for allowing /take transfers
+-- Returns true if the transfer should be allowed due to being a /take, false otherwise.
+local function CheckTakeCondition(fromTeamID, toTeamID)
+	-- Check if sender is allied and has no active human players
+	if Spring.AreTeamsAllied(fromTeamID, toTeamID) then
+		local senderHasActivePlayer = false
+		for _, playerID in ipairs(Spring.GetPlayerList()) do
+			local _, active, spectator, teamID = Spring.GetPlayerInfo(playerID)
+			if active and not spectator and teamID == fromTeamID then
+				senderHasActivePlayer = true
+				break
+			end
+		end
+		if not senderHasActivePlayer then
+			-- This matches the condition where the engine's /take command would trigger GiveEverythingTo
+			return true
+		end
+	end
+	return false -- Not a /take condition
+end
+
+function gadget:AllowUnitTransfer(unitID, unitDefID, fromTeamID, toTeamID, capture)
+	-- 1. Check for clean override flags (e.g., Market)
+	for _, flagName in ipairs(transferOverrideFlags) do
```

Comment:
This is fine by me, I like the implementation of other gadgets signalling the intent to bypass sharing restrictions. Naming maybe up for debate (I'd maybe do "bypass_sharing_restriction_unit_market" or similar but it makes no difference), but looks way less messy than mixing each gadgets responsibility. Very nice :+1:

Reply:
I went with transfer_override_ as the prefix